x-pack/plugins/rule_registry/server/rule_data_plugin_service/resource_installer.ts

```diff
@@ -146,29 +145,22 @@ export class ResourceInstaller {
   }
 
   private async updateIndexMappings(indexInfo: IndexInfo) {
-    const { logger, getClusterClient } = this.options;
-    const clusterClient = await getClusterClient();
+    const { logger } = this.options;
 
-    logger.debug(`Updating mappings of existing concrete indices for ${indexInfo.baseName}`);
+    const aliases = indexInfo.basePattern;
+    const backingIndices = indexInfo.getPatternForBackingIndices();
 
-    const { body: aliasesResponse } = await clusterClient.indices.getAlias({
-      index: indexInfo.basePattern,
-    });
+    logger.debug(`Updating mappings of existing concrete indices for ${indexInfo.baseName}`);
 
-    const writeIndicesAndAliases = Object.entries(aliasesResponse).flatMap(([index, { aliases }]) =>
-      Object.entries(aliases)
-        .filter(([, aliasProperties]) => aliasProperties.is_write_index)
-        .map(([aliasName]) => ({ index, alias: aliasName }))
-    );
+    // Find all concrete indices for all namespaces of the index.
+    const concreteIndices = await this.fetchConcreteIndices(aliases, backingIndices);
+    const concreteWriteIndices = concreteIndices.filter((item) => item.isWriteIndex);
 
-    await Promise.all(
-      writeIndicesAndAliases.map((indexAndAlias) =>
-        this.updateAliasWriteIndexMapping(indexAndAlias)
-      )
-    );
+    // Update mappings of the found write indices.
+    await Promise.all(concreteWriteIndices.map((item) => this.updateAliasWriteIndexMapping(item)));
```

Comment:
This would fail if the index template was removed in the meantime, because it's only ever created when writing fails (which it doesn't when the index exists but the template doesn't). I'm having a hard time estimating the risk. Could it thwart our future index template updates? :thinking:

Reply:
This is a solid point.

It would be good if we could write a fully resilient and idempotent implementation for index templates. On the other hand, this is only possible if the user deletes an index template manually. The question is: do we need to have a fool-proof implementation or we can live with this one that assumes that users must not touch any application resources? I think we should try to find a way to make it fool-proof, but this should not block our progress.

Off the top of my head, I would:

- Call `installNamespacedIndexTemplate` when getting an instance of `IRuleDataWriter`. The namespace becomes known at this point, and we could install the namespaced index template. This would decouple it from the logic of creation of a concrete index and make sure that the template is always there before any attempt to write to the index. It would not handle a race condition where a user could delete the template between `getWriter()` and `writer.bulk()`.
- Integrate the logic of installing templates into `updateIndexMappings`. We should be able to find all the namespaces of the concrete indices; knowing that, we could build and PUT all the index templates for all the namespaces at the setup phase of rule_registry.

Can we address this in https://github.com/elastic/kibana/issues/108941?

Reply:
@marshallmain @xcrzx do you have ideas?

Reply:
> Integrate the logic of installing templates into updateIndexMappings. We should be able to find all the namespaces of the concrete indices; knowing that, we could build and PUT all the index templates for all the namespaces at the setup phase of rule_registry.

++ I'm planning to Implement this logic in https://github.com/elastic/kibana/issues/108941

Reply:
> Call installNamespacedIndexTemplate when getting an instance of IRuleDataWriter

I think that could not be enough. To be fully protected, we'll need:
1. Install index template
2. Update existing index mappings
3. Perform rollover if needed

Reply:
@xcrzx Initially, step 1 happens after steps 2 and 3, because the namespace is not known when Kibana starts, and we can't install an index template for an unknown namespace. Only when we know the namespace and/or start writing, we can create the template and the concrete index if it doesn't exist.

So the suggestion num 1 is probably not enough, because it's true that if the user deletes the template, and the logic triggers rollovers, we'd end up with concrete indices having incorrect mappings and other properties. That's bad. Suggestion num 2 would fix that, I suppose, (and correct me if I misunderstood but you described a sequence of actions that would need to be done for each concrete index in `installIndexLevelResources`) but we're not going to ship https://github.com/elastic/kibana/issues/108941 in 7.15. That sounds like a dilemma 🙂 

I'd do either of this:
- address num 1 in this PR (targeting 7.15) and target num 2 and https://github.com/elastic/kibana/issues/108941 for 7.16
- address num 1, num 2 and https://github.com/elastic/kibana/issues/108941 in 7.15 if a PR for the latter can be done and merged before Aug 31 (BC3)

Which one of the options are we most confident with?

Reply:
I would probably pick either the 1st or 2nd option to implement, but not do both. I don't think we want to have multiple places where we could be updating the index templates when we could do it in a single place since that could make bugs harder to track down.

Personally, I think initializing the namespace resources each time `getWriter()` is called makes the most sense. We would need to let rule writers know that calling `getWriter()` is more expensive in that case, and possibly modify the API to make it more convenient to reuse the writer instead of calling `getWriter().bulk()` every time (brief discussion similar to this topic https://github.com/elastic/kibana/pull/102586#discussion_r657013304).

I see a few advantages to initializing in `getWriter()` and reusing the writer. 
- First, it minimizes the time period where deleting namespace-specific resources could cause failures. It's still possible for someone to delete the index/template/policy in between `getWriter()` and `bulk()` as @banderror mentioned. But in general that's always going to be a risk - if someone is messing with the internal resources of our app, behavior is going to be undefined and failures will occur.
- It allows us to simplify the logic in `bulk()` and assume that the index exists. We wouldn't have to check errors on the response and create the index in certain cases anymore, any errors would simply be thrown.
- Reusing the writer gives us a place to store information across multiple calls to `bulk()`. @spong is looking into adding a "rule execution id" that would be the same for all alerts generated by a particular rule execution. We could create this ID when `getWriter()` is called and automatically insert it into any alerts written by that particular writer.
- We wouldn't need to find and update namespace-specific resources on Kibana startup, since each rule is responsible for checking the status of its namespace and updating the resources if necessary

The downside would be calling `getWriter()` is a bit more expensive, and rule writers would need know to reuse the same writer for the whole execution to get the "rule execution ID" benefit for free.

Reply:
Regarding num 1: the race condition doesn't sound too terrible as it would just fail and retry on the next attempted write, correct?

Regarding the rollover with missing index templates, that should not happen within 7.15.0, because we don't have older indices to roll over. So it should be enough to do this for 7.16? Or do we have other conditions that could cause a rollover?

Reply:
> Regarding num 1: the race condition doesn't sound too terrible for a short-term fix as it would just fail and retry on the next attempted write, correct?

The race condition doesn't look like a big problem. Initialized writers don't live long enough. In Security Solution, they live for a couple of seconds at most (rule execution duration). So if mappings change during that period of time, write could fail, and in the worst case, we could lose a couple of alerts. The next time we call `getWriter()`, the mappings will be updated, and all subsequent writes should succeed.

> Regarding the rollover with missing index templates, that should not happen within 7.15.0, because we don't have older indices to roll over. So it should be enough to do this for 7.16? Or do we have other conditions that could cause a rollover?

Rollovers happen only when index mapping update fails (see bootstrapping logic diagram https://whimsical.com/ruleregistry-index-bootstrapping-Tdd3V9u38mHB6nBTRTJUs7). So theoretically, a rollover could happen if an index template gets modified somehow, and new mappings are incompatible with the old ones. That said, I don't think we should fix that in 7.15.

Reply:
I agree with most of @marshallmain 's points and like the idea of moving both index template installation and mappings upgrade / rollover to `getWriter`. Let's leave this PR as is and address this in a separate PR as part of https://github.com/elastic/kibana/issues/108941. It will probably target 7.16.

> First, it minimizes the time period where deleting namespace-specific resources could cause failures. It's still possible for someone to delete the index/template/policy in between getWriter() and bulk() as @banderror mentioned. But in general that's always going to be a risk - if someone is messing with the internal resources of our app, behavior is going to be undefined and failures will occur.

++

> It allows us to simplify the logic in bulk() and assume that the index exists. We wouldn't have to check errors on the response and create the index in certain cases anymore, any errors would simply be thrown.

++, big simplification

> We could create this ID when getWriter() is called and automatically insert it into any alerts written by that particular writer.

Not sure about that. A writer bound to a namespaces can be 1) shared between multiple rules, 2) live longer than the period between 2 consecutive rule executions. Rule execution ID should be however unique per 1 execution of 1 rule (if I understood correctly). I'd suggest to think about that separately.

> We wouldn't need to find and update namespace-specific resources on Kibana startup, since each rule is responsible for checking the status of its namespace and updating the resources if necessary

++, big simplification

> The downside would be calling getWriter() is a bit more expensive (we could make this a bit clearer by renaming it to initializeWriter())

Short-term, we could rename it to `initializeWriter()`. Longer-term, I think we need to add some caching for writers with a short TTL, so that when a user has 1000+ rules, it doesn't end up with thousands of requests to ES in order to bootstrap resources in the same namespace. I'd say caching should be a separate ticket.

> Regarding num 1: the race condition doesn't sound too terrible for a short-term fix as it would just fail and retry on the next attempted write, correct?

++ agree

Reply:
Added summary of this discussion to https://github.com/elastic/kibana/issues/108941#issuecomment-905506575